docs/source/user_guide/app.rst

```diff
@@ -1479,6 +1479,113 @@ You can reset your spaces to their default state by setting
     # Reset spaces layout in the App
     session.spaces = None
 
+.. _app-saving-spaces-python:
+
+Saving workspaces in Python
+---------------------------
+
+
+If you find yourself frequently using/recreating a certain spaces layout, you
+can save it as a workspace with a name of your choice, using
+:meth:`save_workspace() <fiftyone.core.dataset.Dataset.save_workspace>`.
+For example, to save the workspace defined in the
+:ref:`example above <app-spaces-python>`:
+
+.. code-block:: python
+    :linenos:
+
+    import fiftyone as fo
+    import fiftyone.zoo as foz
+
+    dataset = foz.load_zoo_dataset("quickstart")
+
+    samples_panel = fo.Panel(type="Samples", pinned=True)
+
+    histograms_panel = fo.Panel(
+        type="Histograms",
+        state=dict(plot="Labels"),
+    )
+
+    embeddings_panel = fo.Panel(
+        type="Embeddings",
+        state=dict(brainResult="img_viz", colorByField="metadata.size_bytes"),
+    )
+
+    workspace = fo.Space(
+        children=[
+            fo.Space(
+                children=[
+                    fo.Space(children=[samples_panel]),
+                    fo.Space(children=[histograms_panel]),
+                ],
+                orientation="horizontal",
+            ),
+            fo.Space(children=[embeddings_panel]),
+        ],
+        orientation="vertical",
+    )
+
+    workspace_name = "my-workspace"
+    description = "Samples, embeddings, histograms, oh my!"
+    color = "#FF6D04"
+    dataset.save_workspace(
+        workspace_name,
+        workspace,
+        description=description
+        color=color
+    )
+
+Then in a future session you can load the workspace by name with
+:meth:`load_workspace() <fiftyone.core.dataset.Dataset.load_workspace>`:
+
+.. code-block:: python
+    :linenos:
+
+    import fiftyone as fo
+
+    dataset = fo.load_dataset("quickstart")
+
+    # Retrieve a saved workspace and launch app with it
+    workspace = dataset.load_workspace("my-workspace")
+    session = fo.launch_app(dataset, spaces=workspace)
```

Comment:
@swheaton @benjaminpkane in the case of saved views, the App "knows" whether the current view is named:

```py
view = dataset.load_saved_view(name)
assert view.name == name

session.view = view
session.refresh()

# The session "knows" its a saved view, also reflected in the URL via `?view=name`
assert session.view.name == name
```

This is relevant in many contexts, eg you might want to write an operator to rename the current view and thus `ctx.view` needs to "know" that the current view is named (recently supported by @imanjra in https://github.com/voxel51/fiftyone/pull/4200).

I expect the same will be true for saved workspaces:

```py
workspace = dataset.load_workspace(name)
assert workspace.name == name

session.spaces = workspace
session.refresh()

# The session "knows" its a saved workspace, also reflected in the URL via `?workspace=name`
session.spaces.name == name
```

@swheaton was initially going to propose that we have a `Workspace(Space)` class that adds a `name` property, but I suppose `Space` also needs the `name` property because this would be a natural thing to do:

```py
# Unsaved workspace
assert session.spaces.name is None
```

@benjaminpkane @imanjra for the idea of the App knowing `?workspace=name`

Reply:
ok I don't see an issue with that.

Reply:
Yeah I had considered deprecating `spaces` and adding an alias argument `workspace`, but decided against it, for same reasons

Reply:
Yah, the `Space` will need to have an optional name property to preserve the name in the session.spaces. Alternatively, `session.spaces` can be Union of `Workspace` and `Space` and workspace serialization could have the name property. Either would be fine as long we can preserve the `space.name` (at least for root space) when sending down serialized session.spaces state

Reply:
I think a union of `Workspace` and `Space` makes sense for `session.spaces`. If not, `Spaces` (which are nest-able) has an somewhat unnecessary/misleading attribute

Reply:
This is true. If that's the case, I may need some pointers on implementation.

Reply:
I'm thinking we should just stick with the `Space` class with a nullable `name` property.

If we're going to be strict about backwards-compatibility, we need to still support this:

```py
session.spaces = fo.Space(...)
```

Together with my desire for `session.spaces.name` to always exist going forward, this means we have to add the `name` property to `Space` regardless.

Reply:
@benjaminpkane I have a solution to above that I'm ready to push, passes tests, etc. Can you give me a test that needs to be true in order for session.spaces serialization in app routes to work with name property?

> Either would be fine as long we can preserve the space.name (at least for root space) when sending down serialized session.spaces state

Reply:
The name should be accessible in the serialize spaces via the [StateDescription](https://github.com/voxel51/fiftyone/blob/develop/fiftyone/core/state.py)

Reply:
Ok I'm thinking it should work as long as, say, `from_json(to_json())` round trip works and has name field, which it does.
Is there a typescript definition of statedescription i need to be concerned with?

Reply:
Ibrahim or I can follow up with typescript addition(s)

Reply:
> @benjaminpkane I have a solution to above that I'm ready to push, passes tests, etc.

Once you push, I can start work on the `workspace` URL parameter in earnest

Reply:
sorry didnt realize you were blocked on me for anything. it's pushed now